src/native/http_request_utils.c

```
@@ -347,6 +372,36 @@ struct aws_http_message *aws_http_request_new_from_java_http_request(
     return NULL;
 }
 
+struct aws_http_headers *aws_http_headers_new_from_java_http_headers(JNIEnv *env, jbyteArray marshalled_headers) {
+    const char *exception_message = NULL;
+    struct aws_http_headers *headers = aws_http_headers_new(aws_jni_get_allocator());
+    if (headers == NULL) {
+        aws_jni_throw_runtime_exception(env, "aws_http_headers_new_from_java_http_headers: Unable to allocate headers");
+        return NULL;
+    }
+    const size_t marshalled_headers_length = (*env)->GetArrayLength(env, marshalled_headers);
+
+    jbyte *marshalled_headers_data = (*env)->GetPrimitiveArrayCritical(env, marshalled_headers, NULL);
+    struct aws_byte_cursor marshalled_cur =
+        aws_byte_cursor_from_array((uint8_t *)marshalled_headers_data, marshalled_headers_length);
+    int result = s_unmarshal_http_headers(headers, &marshalled_cur);
+    (*env)->ReleasePrimitiveArrayCritical(env, marshalled_headers, marshalled_headers_data, 0);
+
+    if (result) {
+        exception_message = "aws_http_headers_new_from_java_http_headers: Invalid marshalled headers data.";
```

can we just throw the exception here instead?